analytics/loggable_objects.go

```diff
@@ -0,0 +1,76 @@
+package analytics
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/mxmCherry/openrtb"
+	"net/http"
+)
+
+type RequestType string
+
+const COOKIE_SYNC RequestType = "cookie_sync"
+const AUCTION RequestType = "auction"
+const SETUID RequestType = "set_uid"
+
+//Loggable object of a transaction at /openrtb2/auction endpoint
+type AuctionObject struct {
+	Type               RequestType
+	Status             int
+	Error              []error
+	AdapterBidRequests []LoggableAdapterRequests
+	Request            openrtb.BidRequest
+	Response           openrtb.BidResponse
+	UserAgent          string
+}
+
+//Bid requests from each adapter. This information is available in /openrtb2/auction response but necessary to log in case there are errors
+type LoggableAdapterRequests struct {
+	Name     string
+	Requests string
+	Uri      string
+	Header   http.Header
+	Method   string
+}
+
+//Loggable object of a transaction at /setuid
+type SetUIDObject struct {
+	Type    RequestType
+	Status  int
+	Error   []error
+	Cookie  string
+	Success bool
+}
+
+//Loggable object of a transaction at /cookie_sync
+type CookieSyncObject struct {
+	Type     RequestType
+	Status   int
+	Error    []error
+	Request  string
+	Response string
+}
+
+func (ao AuctionObject) String() string {
+	if content, err := json.Marshal(ao); err != nil {
+		return fmt.Sprintf("Transactional Logs Error: Auction object badly formed %v", err)
+	} else {
+		return string(content)
+	}
+}
+
+func (cso CookieSyncObject) String() string {
+	if content, err := json.Marshal(cso); err != nil {
+		return fmt.Sprintf("Transactional Logs Error: CookieSync object badly formed %v", err)
+	} else {
+		return string(content)
+	}
+}
+
+func (so SetUIDObject) String() string {
```

Comment:
I'm not sure how it works in Golang... but in Java, I've seen a _ton_ of problems come from overriding `toString()` for user-facing features.

The `Stringer` interface (which you're overriding here) is _sort of_ similar to Java's `toString()`. At least, it's similar enough that I don't think we should override it.

I'd recommend defining your own function, like `ToJson() string` (or something like that) instead.